scheduling/taskmetadata.go

```diff
@@ -0,0 +1,533 @@
+	taskRunKey := TaskRunKey{taskID: taskID}
+	taskRunMetadata, err := m.storage.Get(taskRunKey.String())
+	if err != nil {
+		err.AddDetail("task_id", string(taskID))
+		err.AddDetail("run_id", string(runID))
+		return TaskRunMetadata{}, err
+	}
+
+	runs := TaskRuns{}
+	err = runs.Unmarshal([]byte(taskRunMetadata))
+	if err != nil {
+		err.AddDetail("task_id", string(taskID))
+		err.AddDetail("run_id", string(runID))
+		return TaskRunMetadata{}, err
+	}
+
+	// Want to move this logic out
+	found, runRecord := runs.ContainsRun(runID)
+	if !found {
+		err := fferr.NewKeyNotFoundError(taskRunKey.String(), fmt.Errorf("run not found"))
+		err.AddDetail("task_id", string(taskID))
+		err.AddDetail("run_id", string(runID))
+		return TaskRunMetadata{}, err
+	}
+
+	date := runRecord.DateCreated
+	taskRunMetadataKey := TaskRunMetadataKey{taskID: taskID, runID: runRecord.RunID, date: date}
+	rec, err := m.storage.Get(taskRunMetadataKey.String())
+	if err != nil {
+		err.AddDetail("task_id", string(taskID))
+		err.AddDetail("run_id", string(runID))
+		return TaskRunMetadata{}, err
+	}
+
+	taskRun := TaskRunMetadata{}
+	err = taskRun.Unmarshal([]byte(rec))
+	if err != nil {
+		err.AddDetail("task_id", string(taskID))
+		err.AddDetail("run_id", string(runID))
+		return TaskRunMetadata{}, err
+	}
+	return taskRun, nil
+}
+
+func (m *TaskMetadataManager) GetRunsByDate(start time.Time, end time.Time) (TaskRunList, fferr.GRPCError) {
+	// the date range is inclusive
+	var runs []TaskRunMetadata
+
+	// iterate through each day in the date range including the end date
+	for date := start; date.Before(end) || date.Equal(end); date = date.AddDate(0, 0, 1) {
+		dayRuns, err := m.getRunsForDay(date, start, end)
+		if err != nil {
+			return []TaskRunMetadata{}, err
+		}
+		runs = append(runs, dayRuns...)
+	}
+
+	return runs, nil
+}
+
+func (m *TaskMetadataManager) getRunsForDay(date time.Time, start time.Time, end time.Time) ([]TaskRunMetadata, fferr.GRPCError) {
+	key := TaskRunMetadataKey{date: date}
+	recs, err := m.storage.List(key.String())
+	if err != nil {
+		err.AddDetail("date", date.String())
+		return []TaskRunMetadata{}, err
+	}
+
+	var runs []TaskRunMetadata
+	for _, record := range recs {
+		taskRun := TaskRunMetadata{}
+		err = taskRun.Unmarshal([]byte(record))
+		if err != nil {
+			return []TaskRunMetadata{}, err
+		}
+
+		// if the task run started before the start time or after the end time, skip it
+		if taskRun.StartTime.Before(start) || taskRun.StartTime.After(end) {
+			continue
+		}
+		runs = append(runs, taskRun)
+	}
+	return runs, nil
+}
+
+func (m *TaskMetadataManager) GetAllTaskRuns() (TaskRunList, fferr.GRPCError) {
+	recs, err := m.storage.List(TaskRunMetadataKey{}.String())
+	if err != nil {
+		return []TaskRunMetadata{}, err
+	}
+
+	var runs []TaskRunMetadata
+	for _, record := range recs {
+		taskRun := TaskRunMetadata{}
+		err = taskRun.Unmarshal([]byte(record))
```

Comment:
In a lot of situations we are throwing away a lot of these, but are parsing them each time. It's not a problem but maybe leave a comment about it. In case it becomes an issue later.